ImageQualityController thrashes a timer

smfr opened the pull request:
#### 42d62602b5645e08527e41e02f7752c3c73aa373
<pre>
ImageQualityController thrashes a timer
<a href="https://bugs.webkit.org/show_bug.cgi?id=276347">https://bugs.webkit.org/show_bug.cgi?id=276347</a>
<a href="https://rdar.apple.com/131352675">rdar://131352675</a>

Reviewed by Tim Horton.

For content with lots of images, `ImageQualityController::chooseInterpolationQuality()` is called each time
an image paints and potentially restarts a timer. Use a DeferrableOneShotTimer instead, which is designed
for this kind of use case.

Also do some other minor optimizations; only call `context.getCTM().isIdentityOrTranslationOrFlipped()` if we need to,
and only set a new value in the hash tables if the size changed.

* Source/WebCore/rendering/ImageQualityController.cpp:
(WebCore::ImageQualityController::ImageQualityController):
(WebCore::ImageQualityController::restartTimer):
* Source/WebCore/rendering/ImageQualityController.h:

Canonical link: <a href="https://commits.webkit.org/280872@main">https://commits.webkit.org/280872@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/357aea5bc68a8d3f2400d49ff2822f4760b9c630

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57886 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37214 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10362 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61508 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8331 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60014 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44850 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8519 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46910 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5932 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59916 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34903 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50031 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27737 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31670 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7325 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7335 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53618 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7594 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63194 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1800 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7663 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54134 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1806 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50042 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54263 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12805 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1535 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33043 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34129 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35213 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33874 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -4807,6 +4819,7 @@ void LocalFrameView::paintContents(GraphicsContext& context, const IntRect& dirt
     }
 
     didPaintContents(context, dirtyRect, paintingState);
+    didPaintSomeContents();
```
Should `didPaintContents` calls `didPaintSomeContents` instead?

smfr replied:
That would result in double calls for the RenderView's layer.

smfr committed:
ImageQualityController thrashes a timer
https://bugs.webkit.org/show_bug.cgi?id=276347
rdar://131352675

Reviewed by Tim Horton.

For content with lots of images, `ImageQualityController::chooseInterpolationQuality()` is called each time
an image paints and potentially restarts a timer. Use a DeferrableOneShotTimer instead, which is designed
for this kind of use case.

Also do some other minor optimizations; only call `context.getCTM().isIdentityOrTranslationOrFlipped()` if we need to,
and only set a new value in the hash tables if the size changed.

* Source/WebCore/rendering/ImageQualityController.cpp:
(WebCore::ImageQualityController::ImageQualityController):
(WebCore::ImageQualityController::restartTimer):
* Source/WebCore/rendering/ImageQualityController.h:

Canonical link: https://commits.webkit.org/280872@main
